### PR TITLE
Added precompiled modexp (issue #2508)

### LINKED ISSUE #2508: Metropolis-upgrade to chain parameters for the test
https://github.com/ethereum/solidity/blob/aa262d5c8febadd067539c1e05232b1b8c91b4f5/test/RPCSession.cpp#L214 has the chain parameters used for testing Solidity.

Here, only the first four precompiled contracts are listed, but after Metropolis more precompiled contracts should appear.

@chriseth do you agree this should be done?

### PR DESCRIPTION
**Pull request by limexp**
Fixes #2508. Depends on #3117.

### CODE REVIEW

**Comment by axic** [comment]
This shows that our wallet contract test uses precompile addresses:
```
/home/travis/build/ethereum/solidity/test/contracts/Wallet.cpp(563): error in "multisig_value_transfer": check balanceAt(Address(0x05)) == 0 failed [1 != 0]
/home/travis/build/ethereum/solidity/test/contracts/Wallet.cpp(568): error in "multisig_value_transfer": check balanceAt(Address(0x05)) == 0 failed [1 != 0]
/home/travis/build/ethereum/solidity/test/contracts/Wallet.cpp(573): error in "multisig_value_transfer": check balanceAt(Address(0x05)) == 0 failed [1 != 0]
/home/travis/build/ethereum/solidity/test/contracts/Wallet.cpp(579): error in "multisig_value_transfer": check balanceAt(Address(0x05)) == 100 failed [101 != 100]
/home/travis/build/ethereum/solidity/test/contracts/Wallet.cpp(625): error in "revoke_transaction": check balanceAt(Address(0x05)) == 0 failed [1 != 0]
/home/travis/build/ethereum/solidity/test/contracts/Wallet.cpp(630): error in "revoke_transaction": check balanceAt(Address(0x05)) == 0 failed [1 != 0]
/home/travis/build/ethereum/solidity/test/contracts/Wallet.cpp(635): error in "revoke_transaction": check balanceAt(Address(0x05)) == 0 failed [1 != 0]
/home/travis/build/ethereum/solidity/test/contracts/Wallet.cpp(642): error in "revoke_transaction": check balanceAt(Address(0x05)) == 0 failed [1 != 0]
/home/travis/build/ethereum/solidity/test/contracts/Wallet.cpp(648): error in "revoke_transaction": check balanceAt(Address(0x05)) == 100 failed [101 != 100]
/home/travis/build/ethereum/solidity/test/contracts/Wallet.cpp(664): error in "daylimit": check balanceAt(Address(0x05)) == 0 failed [1 != 0]
/home/travis/build/ethereum/solidity/test/contracts/Wallet.cpp(671): error in "daylimit": check balanceAt(Address(0x05)) == 0 failed [1 != 0]
/home/travis/build/ethereum/solidity/test/contracts/Wallet.cpp(680): error in "daylimit": check balanceAt(Address(0x05)) == 0 failed [1 != 0]
/home/travis/build/ethereum/solidity/test/contracts/Wallet.cpp(688): error in "daylimit": check balanceAt(Address(0x05)) == 90 failed [91 != 90]
```

**Comment by axic** [comment]
@gumb0 @pirapira can you please double check the genesis settings here match that of cpp-ethereum?

**Comment by pirapira** [comment]
@axic which chain config should it be?  Mainnet?

**Comment by axic** [comment]
@pirapira just that the precompile address & gas costs match that of mainnet.

**Comment by pirapira** [comment]
@axic it looks different from the mainnet config https://github.com/ethereum/cpp-ethereum/blob/develop/libethashseal/genesis/mainNetwork.cpp

**Comment by axic** [comment]
@pirapira @gumb0 that is actually weird.

From EIPs:
- The gas costs of the precompiled contract are `80 000 * k + 100 000`, where `k` is the number of points or, equivalently, the length of the input divided by 192.
- Gas cost for ``ECADD``: 500
- Gas cost for ``ECMUL``: 40000
- MODEXP: `floor(mult_complexity(max(length_of_MODULUS, length_of_BASE)) * max(ADJUSTED_EXPONENT_LENGTH, 1) / GQUADDIVISOR)` where `GQUADDIVISOR: 20`

This PR:
```
			"0000000000000000000000000000000000000005": { "wei": "1", "precompiled": { "name": "modexp", "modexp": { "divisor": 20 } } },
 			"0000000000000000000000000000000000000006": { "wei": "1", "precompiled": { "name": "alt_bn128_G1_add", "linear": { "base": 500, "word": 0 } } },
 			"0000000000000000000000000000000000000007": { "wei": "1", "precompiled": { "name": "alt_bn128_G1_mul", "linear": { "base": 40000, "word": 0 } } },
 			"0000000000000000000000000000000000000008": { "wei": "1", "precompiled": { "name": "alt_bn128_pairing_product", "alt_bn128_pairing_product": { "base": 100000, "pair": 80000 } } }
```

Here `alt_bn128_pairing_product` seems to be wrong.

Stock cpp-ethereum:
```
		"0000000000000000000000000000000000000005": { "precompiled": { "name": "modexp", "startingBlock": "0x42ae50" } },
		"0000000000000000000000000000000000000006": { "precompiled": { "name": "alt_bn128_G1_add", "startingBlock": "0x42ae50", "linear": { "base": 500, "word": 0 } } },
		"0000000000000000000000000000000000000007": { "precompiled": { "name": "alt_bn128_G1_mul", "startingBlock": "0x42ae50", "linear": { "base": 2000, "word": 0 } } },
		"0000000000000000000000000000000000000008": { "precompiled": { "name": "alt_bn128_pairing_product", "startingBlock": "0x42ae50" } },
```

Here `alt_bn128_G1_mul` seems to be wrong, not sure about `modexp` and `alt_bn128_pairing_product`.

**Comment by axic** [comment]
Looking at the pricers https://github.com/ethereum/cpp-ethereum/blob/d97f2574247b6bf6ab799ad1ea9758228217baf7/libethcore/Precompiled.cpp:

- `modexp` has the divisor hardcoded (no need to specify in the genesis)
- `alt_bn128_pairing_product` has the values hardcoded (no need to specify in the genesis)

File: test/RPCSession.cpp [hunk 1]
```diff
@@ -240,9 +240,10 @@ void RPCSession::test_setChainParams(vector<string> const& _accounts)
 			"0000000000000000000000000000000000000002": { "wei": "1", "precompiled": { "name": "sha256", "linear": { "base": 60, "word": 12 } } },
 			"0000000000000000000000000000000000000003": { "wei": "1", "precompiled": { "name": "ripemd160", "linear": { "base": 600, "word": 120 } } },
 			"0000000000000000000000000000000000000004": { "wei": "1", "precompiled": { "name": "identity", "linear": { "base": 15, "word": 3 } } },
-			"0000000000000000000000000000000000000006": { "wei": "1", "precompiled": { "name": "alt_bn128_G1_add", "linear": { "base": 15, "word": 3 } } },
-			"0000000000000000000000000000000000000007": { "wei": "1", "precompiled": { "name": "alt_bn128_G1_mul", "linear": { "base": 15, "word": 3 } } },
-			"0000000000000000000000000000000000000008": { "wei": "1", "precompiled": { "name": "alt_bn128_pairing_product", "linear": { "base": 15, "word": 3 } } }
+			"0000000000000000000000000000000000000005": { "wei": "1", "precompiled": { "name": "modexp", "modexp": { "divisor": 20 } } },
```
Comment (axic) [inline test/RPCSession.cpp]: Please remove `divisor`.

File: test/RPCSession.cpp [hunk 2]
```diff
@@ -240,9 +240,10 @@ void RPCSession::test_setChainParams(vector<string> const& _accounts)
...
+			"0000000000000000000000000000000000000006": { "wei": "1", "precompiled": { "name": "alt_bn128_G1_add", "linear": { "base": 500, "word": 0 } } },
+			"0000000000000000000000000000000000000007": { "wei": "1", "precompiled": { "name": "alt_bn128_G1_mul", "linear": { "base": 40000, "word": 0 } } },
+			"0000000000000000000000000000000000000008": { "wei": "1", "precompiled": { "name": "alt_bn128_pairing_product", "alt_bn128_pairing_product": { "base": 100000, "pair": 80000 } } }
```
Comment (axic) [inline test/RPCSession.cpp]: Please remove `base` and `pair`.

**Review by axic — APPROVED** [review]

**Pull request merged**